### PR TITLE
adapter: use the new `Optimize` API in `sequence_peek`

### PR DESCRIPTION
**Pull request by aalexandrov**
Add an `OptimizePeek` struct using the new API and integrate it in the peek-related code paths.

### Motivation

  * This PR adds a known-desirable feature.

Continues integrating the `Optimize` API proposed in #20569 according to the rollout plan from the design doc.

### Tips for reviewer

This is a a big diff, but again the first commit is mostly copy-pasting code so we can have the old code path behind a feature flag as a precaution.

* I had to do some ugly special casing in `optimize/peek.rs` in order to avoid changing the `EXPLAIN` output. In the long term this should be removed.
* The PR currently includes a commit that comments out a `TRACE` level `event!` in `sequence_peek_stage`, because with that line enabled I was seeing CI output logs getting in the 100+MB size range.
* I've split the "fork + adapt" commits for `sequence_peek` and `sequence_explain_plan` in two pairs of two, so I can open the `EXPLAIN` changes in a follow-up PR if needed.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/adapter/src/coord/sequencer/inner.rs [hunk 1]
```diff
@@ -2171,7 +2171,7 @@ impl Coordinator {
     ) {
         // Process the current stage and allow for processing the next.
         loop {
-            event!(Level::TRACE, stage = format!("{:?}", stage));
+            // event!(Level::TRACE, stage = format!("{:?}", stage));
```
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner.rs:2174]: I'll have to figure out why `Level::TRACE` is logged in CI and revert this commit before we merge this.

**Review by mjibson — APPROVED** [review]
Phew that's a lot of work. Looks great.

**Review by teskje — APPROVED** [review]
LGTM!

File: src/adapter/src/optimize/peek.rs [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+impl Debug for GlobalMirPlan<Unresolved> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GlobalMirPlan")
+            .field("df_desc", &self.df_desc)
+            .field("df_meta", &self.df_meta)
+            .finish()
+    }
+}
+
+impl<'s> Clone for GlobalMirPlan<ResolvedGlobal<'s>> {
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/adapter/src/optimize/peek.rs:115]: It didn't in an earlier version, but now it does.

File: src/adapter/src/optimize/mod.rs [hunk 3]
```diff
@@ -50,14 +50,16 @@
...
+pub mod subscribe;
+pub mod view;
 
 // Re-export optimzier structs
-pub use index::{Index, OptimizeIndex};
+pub use index::OptimizeIndex;
 pub use materialized_view::OptimizeMaterializedView;
+pub use peek::OptimizePeek;
 pub use subscribe::OptimizeSubscribe;
 pub use view::OptimizeView;
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/adapter/src/optimize/mod.rs:64]: Initially I thought that the only types that are needed are the `Optimize$T` types, as in general the pipeline starts from some public type (like `MirRelationExpr`) and the module-specific types that representing intermediate results usually don't need to be imported in client modules.

However, some `Optimize` implementations need a bespoke type to start (I think that `Index` is the only exception to that rule at the moment). If I can't revert to some public `Index` type I'll probably implement your suggestion in one of the next PRs.

[the remark this answers is not in the document]
Comment (teskje) [reply src/adapter/src/optimize/mod.rs:66]: The as-of selection fix will also need to name the intermediary types, to be able to return optimized but unresolved plans from functions and store them, so I think they (or at least some of them) will need to be accessible anyway.

[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/adapter/src/optimize/mod.rs:66]: After (and before) the change the line looks like that:

```rust
// MIR ⇒ MIR optimization (global)
let global_mir_plan = optimizer.optimize(from)?;
// Timestamp selection
let as_of = ...
let global_mir_plan = global_mir_plan.resolve(Antichain::from_elem(as_of));
// MIR ⇒ LIR lowering and LIR ⇒ LIR optimization (global)
let global_lir_plan = optimizer.optimize(global_mir_plan.clone())?;
```

[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/adapter/src/optimize/mod.rs:66]: I would like to merge this now in order to resolve conflicts. I will add a commit that fixes the naming in #22971.

[the remark this answers is not in the document]
Comment (teskje) [reply src/adapter/src/optimize/mod.rs:66]: Fine with me!

**Pull request merged**